Geometry/TrackerCommonData/plugins/dd4hep/DDTrackerRingAlgo.cc

```diff
@@ -0,0 +1,129 @@
+			  << convertRadToDeg(rangeAngle) << " " << convertRadToDeg(delta)
+			  << " Radius " << radius << " Centre " << center[0]
+			  << ", " << center[1] << ", "<< center[2];
+
+  LogDebug("TrackerGeom") << "DDTrackerRingAlgo debug: Parent " << parentName
+			  << "\tChild " << childName << " NameSpace "
+			  << ns.name();
+
+  Rotation3D flipRot, tiltRot, phiRot, globalRot; // Identity
+  Rotation3D flipMatrix, tiltMatrix, phiRotMatrix, globalRotMatrix; // Identity matrix
+  
+  // flipMatrix calculus
+  if(isFlipped) {
+    LogDebug("TrackerGeom") << "DDTrackerRingAlgo test: Creating a new rotation: "
+			    << "\t90., 180., "
+			    << "90., 90., "
+			    << "180., 0.";
+    flipRot = makeRotation3D(90._deg, 180._deg, 90._deg, 90._deg, 180._deg, 0._deg);
+    flipMatrix = flipRot;
+  }
+  // tiltMatrix calculus
+  if(isZPlus) {
+    LogDebug("TrackerGeom") << "DDTrackerRingAlgo test: Creating a new rotation: "
+			    << "\t90., 90., "
+			    << convertRadToDeg(tiltAngle) << ", 180., "
+			    << convertRadToDeg(90._deg - tiltAngle) << ", 0.";
+    tiltRot = makeRotation3D(90._deg, 90._deg, tiltAngle, 180._deg, 90._deg - tiltAngle, 0._deg);
+    tiltMatrix = tiltRot;
+    if(isFlipped) { tiltMatrix *= flipMatrix; }
+  }
+  else {
+    LogDebug("TrackerGeom") << "DDTrackerRingAlgo test: Creating a new rotation: "
+			    << "\t90., 90., "
+			    << convertRadToDeg(tiltAngle) << ", 0., "
+			    << convertRadToDeg(90._deg + tiltAngle) << ", 0.";
+    tiltRot = makeRotation3D(90._deg, 90._deg, tiltAngle, 0._deg, 90._deg + tiltAngle, 0._deg);
+    tiltMatrix = tiltRot;
+    if(isFlipped) { tiltMatrix *= flipMatrix; }
+  }
+  
+  // Loops for all phi values
+  double theta  = 90._deg;
+  int    copy   = startCopyNo;
+  double phi    = startAngle;
+  
+  for(int i = 0; i < n; ++i) {
+    // phiRotMatrix calculus
+    double phix = phi;
+    double phiy = phix + 90._deg;
+    double phideg = convertRadToDeg(phix);  
+    if(phideg != 0.) {
+      LogDebug("TrackerGeom") << "DDTrackerRingAlgo test: Creating a new rotation: "
+			      << "\t90., " << convertRadToDeg(phix)
+			      << ", 90.," << convertRadToDeg(phiy)
+			      <<", 0., 0.";
+      phiRot = makeRotation3D(theta, phix, theta, phiy, 0., 0.);
+      phiRotMatrix = phiRot;
+    }
+ 
+    // globalRot def
+    globalRotMatrix = phiRotMatrix * tiltMatrix;
+    globalRot = Rotation3D(globalRotMatrix);
+    
+    // translation def
+    double xpos = radius*cos(phi) + center[0];
+    double ypos = radius*sin(phi) + center[1];
+    double zpos = center[2];
+    Position tran(xpos, ypos, zpos);
+   
+    // Positions child with respect to parent
+    mother.placeVolume(child, copy, Transform3D(globalRot, tran));
+    LogDebug("TrackerGeom") << "DDTrackerRingAlgo test " << child << " number "
+			    << copy << " positioned in " << mother << " at "
+			    << tran  << " with " << globalRot;
+    
+    copy += incrCopyNo;
+    phi  += delta;
+  }
+  return 1;
```

Comment:
what does this return value denote?

Reply:
1 means that the algorithm was executed.